pkg/services/ngalert/provisioning/notification_policies.go

```diff
@@ -154,3 +173,117 @@ func (nps *NotificationPolicyService) ensureDefaultReceiverExists(cfg *definitio
 	nps.log.Error("Grafana Alerting has been configured with a default configuration that is internally inconsistent! The default configuration's notification policy must have a corresponding receiver.")
 	return fmt.Errorf("inconsistent default configuration")
 }
+
+func calculateRouteFingerprint(route definitions.Route) (string, error) {
+	sum := fnv.New64a()
+	err := writeToHash(sum, &route)
+	if err != nil {
+		return "", fmt.Errorf("failed to calculate fingerprint: %w", err)
+	}
+	return fmt.Sprintf("%016x", sum.Sum64()), nil
+}
+
+func writeToHash(sum hash.Hash, r *definitions.Route) error {
+	writeBytes := func(b []byte) {
+		_, _ = sum.Write(b)
+		// add a byte sequence that cannot happen in UTF-8 strings.
+		_, _ = sum.Write([]byte{255})
+	}
+	writeString := func(s string) {
+		if len(s) == 0 {
+			writeBytes(nil)
+			return
+		}
+		// #nosec G103
+		// avoid allocation when converting string to byte slice
+		writeBytes(unsafe.Slice(unsafe.StringData(s), len(s)))
+	}
+
+	// this temp slice is used to convert ints to bytes.
+	tmp := make([]byte, 8)
+	writeInt := func(u int64) {
+		binary.LittleEndian.PutUint64(tmp, uint64(u))
+		writeBytes(tmp)
+	}
+	writeBool := func(b bool) {
+		if b {
+			writeInt(1)
+		} else {
+			writeInt(0)
+		}
+	}
+	writeDuration := func(d *model.Duration) {
```

Comment:
Should we use a writeDuration function similar to notification settings to differentiate nils?

```go
	writeDuration := func(d *model.Duration) {
		if d == nil {
			_, _ = h.Write([]byte{255})
		} else {
			binary.LittleEndian.PutUint64(tmp, uint64(*d))
			_, _ = h.Write(tmp)
			_, _ = h.Write([]byte{255})
		}
	}
```

Reply:
+1